File: source/bxgenbb_help/src/decay0/gfang.cc

#include <cmath>

#include <gsl/gsl_math.h>

#include <mygsl/rng.h>

#include <genbb_help/primary_event.h>

#include <genbb_help/decay0/gfang.h>

namespace genbb {
  namespace decay0 {
  
    void decay0_gfang(const double * p, 
                      double &costh, double &sinth, double &cosph, double &sinph, 
                      bool &rotate)
    {
      /*
        // Could also use CLHEP 3D-vector class :
        geomtools::vector_3d pp(p[0],p[1],p[2]);
        std::clog << "Cos(theta) = " << std::cos(pp.theta()) 
        << " (" << pp.cosTheta() << ')' << '\n';
        std::clog << "Sin(theta) = " << std::sin(pp.theta()) << '\n';
        std::clog << "Cos(phi)   = " << std::cos(pp.phi()) << '\n';
        std::clog << "Sin(phi)   = " << std::sin(pp.phi()) << '\n';
        std::clog << '\n';
      */
      static const double one = 1.0;
      // additional (as compared with original GFANG) normalization
      double pfull=std::sqrt(gsl_pow_2(p[0])
                             +gsl_pow_2(p[1])
                             +gsl_pow_2(p[2]));
      double dux=p[0]/pfull;
      double duy=p[1]/pfull;
      double duz=p[2]/pfull;
      rotate=true;
      if (std::abs(duz) >= 0.85) {
        double dsith2=gsl_pow_2(dux)+gsl_pow_2(duy);
        if(dsith2 > 0.) {
          double sduz = (duz > 0)? one: -one;
          costh=sduz*std::sqrt(one-dsith2);
          double dsith=std::sqrt(dsith2);
          sinth=dsith;
          cosph=dux/dsith;
          sinph=duy/dsith;
        }
        else if (duz > 0.) {
          rotate=false;
          costh=1.;
          sinth=0.;
          cosph=1.;
          sinph=0.;
        }
        else {
          costh=-1.;
          sinth=0.;
          cosph=1.;
          sinph=0.;
        }
      }
      else {
        costh=duz;
        double dsith=std::sqrt((one+duz)*(one-duz));
        sinth=dsith;
        double dnorm=one/std::sqrt(gsl_pow_2(dux)+gsl_pow_2(duy));
        cosph=dux*dnorm;
        sinph=duy*dnorm;
      }
      return;
    }
    
  } // end of namespace decay0 
} // end of namespace genbb 

File: source/bxgenbb_help/src/decay0_fortran/gfang.f
c gfang.f 
c
c This file was extracted from the 'decay0' program by V.I. Tretyak
c
c Copyright 1995-2011 V.I. Tretyak
c
c This program is free software; you can redistribute it and/or modify
c it under the terms of the GNU General Public License as published by
c the Free Software Foundation; either version 3 of the License, or (at
c your option) any later version.
c 
c This program is distributed in the hope that it will be useful, but
c WITHOUT ANY WARRANTY; without even the implied warranty of
c MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
c General Public License for more details.
c 
c You should have received a copy of the GNU General Public License
c along with this program; if not, write to the Free Software
c Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
c 
c
	subroutine gfang(p,costh,sinth,cosph,sinph,rotate)
c
c Finds theta-phi angles for a particle with momentum P.
c The logical flag ROTATE indicates whether the director
c cosines are along the Z axis (.FALSE.) or not (.TRUE.).
c
c F.Carminati (GEANT).
c
	dimension p(3)
	logical rotate
	parameter (one=1)
c additional (as compared with original GFANG) normalization
	pfull=sqrt(p(1)**2+p(2)**2+p(3)**2)
	dux=p(1)/pfull
	duy=p(2)/pfull
	duz=p(3)/pfull
	rotate=.true.
	if(abs(duz).ge.0.85) then
	   dsith2=dux**2+duy**2
	   if(dsith2.gt.0.) then
	      costh=sign(one,duz)*sqrt(one-dsith2)
	      dsith=sqrt(dsith2)
	      sinth=dsith
	      cosph=dux/dsith
	      sinph=duy/dsith
	   elseif(duz.gt.0.) then
	      rotate=.false.
	      costh=1.
	      sinth=0.
	      cosph=1.
	      sinph=0.
	   else
	      costh=-1.
	      sinth=0.
	      cosph=1.
	      sinph=0.
	   endif
	else
	   costh=duz
	   dsith=sqrt((one+duz)*(one-duz))
	   sinth=dsith
	   dnorm=one/sqrt(dux**2+duy**2)
	   cosph=dux*dnorm
	   sinph=duy*dnorm
	endif
	return
	end
c
c end of gfang.f
c
c Local Variables: --
c mode: fortran --
c End: --
